File: experimental/torch_ops_extension/custom_ops/converter/npu_fused_infer_attention_score.py
from typing import (
    Any, Callable, ContextManager, Iterator, List, Literal, NamedTuple, Optional, Sequence, Tuple, TypeVar,
    Union, overload,
)
import torch
import torch_npu
import torchair
from torch import Generator, contiguous_format, inf, strided, SymInt
from torch.types import Device, Number, _bool, _complex, _device, _dtype, _float, _int, _layout, _qscheme, _size
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import declare_supported, register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec, DataType
from torchair._ge_concrete_graph.supported_declaration import _TypedTensor, F32, F16, F64, I32, I16, I64, I8, U8, \
    BOOL, Support
from torchair._ge_concrete_graph.utils import dtype_promote
from torchair.ge import attr


# 为自定义算子注册converter，用于torch.compile 场景成图
@register_fx_node_ge_converter(torch.ops.custom.npu_fused_infer_attention_score.default)
def convert_npu_fused_infer_attention_score(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    *,
    query_rope: Optional[Tensor] = None,
    key_rope: Optional[Tensor] = None,
    pse_shift: Optional[Tensor] = None,
    atten_mask: Optional[Tensor] = None,
    actual_seq_qlen: Optional[Union[List[int], Tensor]] = None,
    actual_seq_kvlen: Optional[Union[List[int], Tensor]] = None,
    block_table: Optional[Tensor] = None,
    dequant_scale_query: Optional[Tensor] = None,
    dequant_scale_key: Optional[Tensor] = None,
    dequant_offset_key: Optional[Tensor] = None,
    dequant_scale_value: Optional[Tensor] = None,
    dequant_offset_value: Optional[Tensor] = None,
    dequant_scale_key_rope: Optional[Tensor] = None,
    quant_scale_out: Optional[Tensor] = None,
    quant_offset_out: Optional[Tensor] = None,
    learnable_sink: Optional[Tensor] = None,
    num_query_heads: int = 1,
    num_key_value_heads: int = 0,
    softmax_scale: float = 1.0,
    pre_tokens: int = 2147483647,
    next_tokens: int = 2147483647,
    input_layout: str = "BSH",
    sparse_mode: int = 0,
    block_size: int = 0,
    query_quant_mode: int = 0,
    key_quant_mode: int = 0,
    value_quant_mode: int = 0,
    inner_precise: int = 0,
    return_softmax_lse: bool = False,
    query_dtype: Optional[int] = None,
    key_dtype: Optional[int] = None,
    value_dtype: Optional[int] = None,
    query_rope_dtype: Optional[int] = None,
    key_rope_dtype: Optional[int] = None,
    key_shared_prefix_dtype: Optional[int] = None,
    value_shared_prefix_dtype: Optional[int] = None,
    dequant_scale_query_dtype: Optional[int] = None,
    dequant_scale_key_dtype: Optional[int] = None,
    dequant_scale_value_dtype: Optional[int] = None,
    dequant_scale_key_rope_dtype: Optional[int] = None,
    out_dtype: Optional[int] = None,
    meta_outputs: TensorSpec = None,):

    print("convert_npu_fused_infer_attention_score")
    def to_ge_int_tensor(x, name):
        if x is None:
            return None
        # x 是 list[int]
        vals = [int(v) for v in x]
        # 下面函数名可能是 Const / const / make_const，请按你torchair版本替换
        return torchair.ge.Const(vals, dtype=DataType.DT_INT64, node_name=name)
    if isinstance(actual_seq_qlen, list):
        actual_seq_qlen = to_ge_int_tensor(actual_seq_qlen, "actual_seq_qlen_const")
    if isinstance(actual_seq_kvlen, list):
        actual_seq_kvlen = to_ge_int_tensor(actual_seq_kvlen, "actual_seq_kvlen_const")

    # dropped params
    quant_scale1 = None
    dequant_scale2 = None
    dequant_scale1 = None
    antiquant_scale = None
    antiquant_offset = None
    query_padding_size = None
    kv_padding_size = None
    key_shared_prefix = None
    value_shared_prefix = None
    actual_shared_prefix_len = None
    q_start_idx=None
    kv_start_idx=None
    # 1) Tensor inputs only
    inputs = {
        "query": query,
        "key": [key],
        "value": [value],
        "pse_shift": pse_shift,
        "atten_mask": atten_mask,
        "actual_seq_lengths": actual_seq_qlen,
        "actual_seq_lengths_kv": actual_seq_kvlen,
        "dequant_scale1": dequant_scale1,
        "quant_scale1": quant_scale1,
        "dequant_scale2": dequant_scale2,
        "quant_scale2": quant_scale_out,
        "quant_offset2": quant_offset_out,
        "antiquant_scale": antiquant_scale,
        "antiquant_offset": antiquant_offset,
        "block_table": block_table,
        "query_padding_size": query_padding_size,
        "kv_padding_size": kv_padding_size,
        "key_antiquant_scale": dequant_scale_key,
        "key_antiquant_offset": dequant_offset_key,
        "value_antiquant_scale": dequant_scale_value,
        "value_antiquant_offset": dequant_offset_value,
        "key_shared_prefix": key_shared_prefix,
        "value_shared_prefix": value_shared_prefix,
        "actual_shared_prefix_len": actual_shared_prefix_len,
        "query_rope": query_rope,
        "key_rope": key_rope,
        "key_rope_antiquant_scale": dequant_scale_key_rope,
        "dequant_scale_query": dequant_scale_query,
        "learnable_sink": learnable_sink,
        "q_start_idx": q_start_idx,
        "kv_start_idx": kv_start_idx,      
    }

    # inputs = {k: v for k, v in inputs.items() if v is not None}
    # 2) Required attrs
    attrs = {
        "num_heads": attr.Int(num_query_heads),
        "scale": attr.Float(softmax_scale),
        "pre_tokens": attr.Int(pre_tokens),
        "next_tokens": attr.Int(next_tokens),
        "input_layout": attr.Str(input_layout),
        "num_key_value_heads": attr.Int(num_key_value_heads),
        "sparse_mode": attr.Int(sparse_mode),
        "inner_precise": attr.Int(inner_precise),
        "block_size": attr.Int(block_size),
        "antiquant_mode": attr.Int(0),
        "softmax_lse_flag": attr.Bool(return_softmax_lse),
        "query_quant_mode": attr.Int(query_quant_mode),
        "key_quant_mode": attr.Int(key_quant_mode),
        "value_quant_mode": attr.Int(value_quant_mode),
        "key_antiquant_mode": attr.Int(0),
        "value_antiquant_mode": attr.Int(0),
        "query_quant_mode": attr.Int(0),
        "pse_type": attr.Int(0),
        "out_dtype": attr.Int(0),
    }
    # # 3) Optional int attrs (only when not None)
    # def add_opt_int(name, v):
    #     if v is not None:
    #         attrs[name] = attr.Int(int(v))
    # add_opt_int("query_dtype", query_dtype)
    # add_opt_int("key_dtype", key_dtype)
    # add_opt_int("value_dtype", value_dtype)
    # add_opt_int("query_rope_dtype", query_rope_dtype)
    # add_opt_int("key_rope_dtype", key_rope_dtype)
    # add_opt_int("key_shared_prefix_dtype", key_shared_prefix_dtype)
    # add_opt_int("value_shared_prefix_dtype", value_shared_prefix_dtype)
    # add_opt_int("dequant_scale_query_dtype", dequant_scale_query_dtype)
    # add_opt_int("dequant_scale_key_dtype", dequant_scale_key_dtype)
    # add_opt_int("dequant_scale_value_dtype", dequant_scale_value_dtype)
    # add_opt_int("dequant_scale_key_rope_dtype", dequant_scale_key_rope_dtype)

    print("end convert_npu_fused_infer_attention_score")
    return torchair.ge.custom_op(
        "FusedInferAttentionScore",
        inputs=inputs,
        attrs=attrs,
        outputs=["attention_out", "softmax_lse"],
    )

File: experimental/ascend_ops/src/csrc/incre_flash_attention/npu_fused_infer_attention_score.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <iostream>
#include <vector>
#include <torch/library.h>
#include <ATen/Operators.h>
#include "torch_npu/csrc/framework/utils/OpPreparation.h"
#include "torch_npu/csrc/framework/OpCommand.h"
#include "op_host/incre_flash_attention_tiling_impl.h"
#include "op_kernel/incre_flash_attention_arch32.h"
namespace custom {

#define LAUNCH_INCRE_FA(FD, LAYOUT, ANTIQ)                                                                 \
    incre_flash_attention<FD, LAYOUT, ANTIQ><<<blockDim, nullptr, aclstream>>>(                            \
        (GM_ADDR)(query.data_ptr()),                                                                       \
        (GM_ADDR)(key.data_ptr()),                                                                         \
        (GM_ADDR)(value.data_ptr()),                                                                       \
        (GM_ADDR)(pse_shift.has_value() ? pse_shift->data_ptr() : nullptr),                                \
        (GM_ADDR)(atten_mask.has_value() ? atten_mask->data_ptr() : nullptr),                              \
        (GM_ADDR)(actual_seq_qlen.has_value() ? actual_seq_qlen->data_ptr() : nullptr),                     \
        (GM_ADDR)(actual_seq_kvlen.has_value() ? actual_seq_kvlen->data_ptr() : nullptr),                   \
        (GM_ADDR)(nullptr),  /* deqScale1 */                                                               \
        (GM_ADDR)(nullptr),  /* quantScale1 */                                                             \
        (GM_ADDR)(nullptr),  /* deqScale2 */                                                               \
        (GM_ADDR)(nullptr),  /* quantScale2 */                                                             \
        (GM_ADDR)(nullptr),  /* quantOffset2 */                                                            \
        (GM_ADDR)(nullptr),  /* antiquantScale */                                                          \
        (GM_ADDR)(nullptr),  /* antiquantOffset */                                                         \
        (GM_ADDR)(block_table.has_value() ? block_table->data_ptr() : nullptr),                            \
        (GM_ADDR)(nullptr),  /* queryPaddingSize */                                                        \
        (GM_ADDR)(nullptr),  /* kvPaddingSize */                                                           \
        (GM_ADDR)(dequant_scale_key.has_value() ? dequant_scale_key->data_ptr() : nullptr),                \
        (GM_ADDR)(nullptr),  /* keyAntiquantOffset */                                                      \
        (GM_ADDR)(dequant_scale_value.has_value() ? dequant_scale_value->data_ptr() : nullptr),            \
        (GM_ADDR)(nullptr),  /* valueAntiquantOffset */                                                    \
        (GM_ADDR)(nullptr),  /* keySharedPrefix */                                                         \
        (GM_ADDR)(nullptr),  /* valueSharedPrefix */                                                       \
        (GM_ADDR)(nullptr),  /* actualSharedPrefixLen */                                                   \
        (GM_ADDR)(query_rope.has_value() ? query_rope->data_ptr() : nullptr),                              \
        (GM_ADDR)(key_rope.has_value() ? key_rope->data_ptr() : nullptr),                                  \
        (GM_ADDR)(dequant_scale_key_rope.has_value() ? dequant_scale_key_rope->data_ptr() : nullptr),      \
        (GM_ADDR)(dequant_scale_query.has_value() ? dequant_scale_query->data_ptr() : nullptr),            \
        (GM_ADDR)(metadata.has_value() ? metadata->data_ptr() : nullptr),            \
        (GM_ADDR)(output.data_ptr()),                                                                      \
        (GM_ADDR)(softmax_lse.data_ptr()),                                                                 \
        (GM_ADDR)(workspaceTensor.data_ptr()),                                                                        \
        tilingData)

const static int FLASH_THRESHOLD = 512;
const static int64_t PFA_SPARSE_HIGH_PRECISION_NO_MASK = 10;
const static int64_t PFA_SPARSE_HIGH_PRECISION_BAND = 14;
const static int64_t DIM_0 = 0;
const static int64_t DIM_1 = 1;
const static int64_t DIM_2 = 2;
const static int64_t DIM_3 = 3;
const static int64_t DIM_4 = 4;
const static int64_t PA_BBH_DIMS = 3;
const static int64_t PA_BNBD_DIMS = 4;
const static int64_t PA_NZ_DIMS = 5;
using namespace at_npu::native;
using namespace optiling;
using npu_preparation = at_npu::native::OpPreparation;

std::tuple<at::Tensor, at::Tensor>
construct_fia_output_tensor_v2(const at::Tensor &query, const at::Tensor &value, std::string input_layout_str,
                               const c10::optional<at::Tensor> &quant_scale_out,
                               const c10::optional<at::Tensor> &block_table, int64_t num_query_heads,
                               int64_t num_key_value_heads, // 增加kvhead用于计算BBH情况下的D
                               bool return_softmax_lse, const c10::optional<at::Tensor> &query_rope)
{
    at::Tensor output;
    int64_t batchSize = 1;
    int64_t qsSize = 1;
    at::Tensor tmp_output = at::empty_like(query);
    if (input_layout_str == "BNSD_BSND") {
        tmp_output = at::empty({query.size(DIM_0), query.size(DIM_2), query.size(DIM_1), query.size(DIM_3)},
                               query.options().dtype(query.dtype()));
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_2);
    } else if (input_layout_str == "BNSD_NBSD") {
        tmp_output = at::empty({query.size(DIM_1), query.size(DIM_0), query.size(DIM_2), query.size(DIM_3)},
                               query.options().dtype(query.dtype()));
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_2);
    } else if (input_layout_str == "BSND_NBSD") {
        tmp_output = at::empty({query.size(DIM_2), query.size(DIM_0), query.size(DIM_1), query.size(DIM_3)},
                               query.options().dtype(query.dtype()));
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_1);
    } else if (input_layout_str == "TND_NTD") {
        tmp_output =
            at::empty({query.size(DIM_1), query.size(DIM_0), query.size(DIM_2)}, query.options().dtype(query.dtype()));
    } else if (input_layout_str == "NSD") {
        batchSize = 1;
        qsSize = query.size(DIM_1);
    } else if (input_layout_str == "BSH") {
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_1);
    } else if (input_layout_str == "BSND") {
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_1);
    } else if (input_layout_str == "BNSD") {
        batchSize = query.size(DIM_0);
        qsSize = query.size(DIM_2);
    } 

    if (quant_scale_out.has_value()) {
        output = at::empty(tmp_output.sizes(), c10::dtype(c10::ScalarType::Char));
    } else if (query.dtype() == at::kChar) {
        if (query_rope.has_value()) {
            const at::Tensor &query_rope_tensor = c10::value_or_else(query_rope, [] { return at::Tensor(); });
            output = at::empty(tmp_output.sizes(), c10::dtype(query_rope_tensor.dtype()));
        } else {
            output = at::empty(tmp_output.sizes(), c10::dtype(c10::ScalarType::Half));
        }
    } else {
        output = at::empty_like(tmp_output);
    }

    auto lse_opts = output.options().dtype(c10::ScalarType::Float);
    at::Tensor softmax_lse;
    softmax_lse = at::empty({batchSize, num_query_heads, qsSize, 1}, lse_opts);

    if (!return_softmax_lse) {
        softmax_lse = at::empty({0}, lse_opts);
    }
    return std::tuple<at::Tensor, at::Tensor>(output, softmax_lse);
}

RequiredParaInfo ToRequiredParaInfo(const at::Tensor &t)
{
    return RequiredParaInfo{.data = t.defined() ? t.data_ptr() : nullptr, .shape = t.sizes(), .dType = t.scalar_type()};
}

OptionalTensorParaInfo ToOptionalTensorParaInfo(const c10::optional<at::Tensor> &ot)
{
    OptionalTensorParaInfo info;
    info.hasValue = ot.has_value();
    if (info.hasValue && ot->defined()) {
        info.data = ot->data_ptr();
        info.shape = ot->sizes();
        info.dType = ot->scalar_type();
    } else {
        info.data = nullptr;
    }
    return info;
}

void ConvertContextToParamsIFA(
    IFAContext &ifaContext, const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const c10::optional<at::Tensor> &query_rope, const c10::optional<at::Tensor> &key_rope,
    const c10::optional<at::Tensor> &pse_shift, const c10::optional<at::Tensor> &atten_mask,
    const c10::optional<at::Tensor> &actual_seq_qlen, const c10::optional<at::Tensor> &actual_seq_kvlen,
    const c10::optional<at::Tensor> &block_table, const c10::optional<at::Tensor> &dequant_scale_query,
    const c10::optional<at::Tensor> &dequant_scale_key, const c10::optional<at::Tensor> &dequant_offset_key,
    const c10::optional<at::Tensor> &dequant_scale_value, const c10::optional<at::Tensor> &dequant_offset_value,
    const c10::optional<at::Tensor> &dequant_scale_key_rope, const c10::optional<at::Tensor> &quant_scale_out,
    const c10::optional<at::Tensor> &quant_offset_out, const c10::optional<at::Tensor> &learnable_sink,
    int64_t num_query_heads, int64_t num_key_value_heads, double softmax_scale, int64_t pre_tokens, int64_t next_tokens,
    const char *input_layout, int64_t sparse_mode, int64_t block_size, int64_t query_quant_mode, int64_t key_quant_mode,
    int64_t value_quant_mode, int64_t inner_precise, bool return_softmax_lse, const c10::optional<int64_t> &query_dtype,
    const c10::optional<int64_t> &key_dtype, const c10::optional<int64_t> &value_dtype,
    const c10::optional<int64_t> &query_rope_dtype, const c10::optional<int64_t> &key_rope_dtype,
    const c10::optional<int64_t> &key_shared_prefix_dtype, const c10::optional<int64_t> &value_shared_prefix_dtype,
    const c10::optional<int64_t> &dequant_scale_query_dtype, const c10::optional<int64_t> &dequant_scale_key_dtype,
    const c10::optional<int64_t> &dequant_scale_value_dtype, const c10::optional<int64_t> &dequant_scale_key_rope_dtype,
    const at::Tensor &attenOut, const at::Tensor &lseOut)
{
    // required input
    ifaContext.opName = "FusedInferAttentionScore";
    ifaContext.query = ToRequiredParaInfo(query);
    ifaContext.key = ToRequiredParaInfo(key);
    ifaContext.value = ToRequiredParaInfo(value);
    // optional input
    ifaContext.pseShift = ToOptionalTensorParaInfo(pse_shift);                           
    ifaContext.attenMask = ToOptionalTensorParaInfo(atten_mask);                         
    ifaContext.actualSeqLengthsQ = ToOptionalTensorParaInfo(actual_seq_qlen);             
    ifaContext.actualSeqLengths = ToOptionalTensorParaInfo(actual_seq_kvlen);             
    ifaContext.deqScale1 = ToOptionalTensorParaInfo(c10::nullopt);                       
    ifaContext.quantScale1 = ToOptionalTensorParaInfo(c10::nullopt);                     
    ifaContext.deqScale2 = ToOptionalTensorParaInfo(c10::nullopt);                       
    ifaContext.quantScale2 = ToOptionalTensorParaInfo(quant_scale_out);                  
    ifaContext.quantOffset2 = ToOptionalTensorParaInfo(quant_offset_out);                
    ifaContext.antiquantScale = ToOptionalTensorParaInfo(c10::nullopt);                  
    ifaContext.antiquantOffset = ToOptionalTensorParaInfo(c10::nullopt);                 
    ifaContext.blockTable = ToOptionalTensorParaInfo(block_table);                       
    ifaContext.queryPaddingSize = ToOptionalTensorParaInfo(c10::nullopt);                
    ifaContext.kvPaddingSize = ToOptionalTensorParaInfo(c10::nullopt);                   
    ifaContext.keyAntiquantScale = ToOptionalTensorParaInfo(dequant_scale_key);          
    ifaContext.keyAntiquantOffset = ToOptionalTensorParaInfo(dequant_offset_key);        
    ifaContext.valueAntiquantScale = ToOptionalTensorParaInfo(dequant_scale_value);      
    ifaContext.valueAntiquantOffset = ToOptionalTensorParaInfo(dequant_offset_value);    
    ifaContext.keySharedPrefix = ToOptionalTensorParaInfo(c10::nullopt);                 
    ifaContext.valueSharedPrefix = ToOptionalTensorParaInfo(c10::nullopt);               
    ifaContext.actualSharedPrefixLen = ToOptionalTensorParaInfo(c10::nullopt);            
    ifaContext.queryRope = ToOptionalTensorParaInfo(query_rope);                         
    ifaContext.keyRope = ToOptionalTensorParaInfo(key_rope);                             
    ifaContext.keyRopeAntiquantScale = ToOptionalTensorParaInfo(dequant_scale_key_rope); 
    ifaContext.dequantScaleQuery = ToOptionalTensorParaInfo(dequant_scale_query);        
    ifaContext.qStartIdx = ToOptionalTensorParaInfo(c10::nullopt);                       
    ifaContext.kvStartIdx = ToOptionalTensorParaInfo(c10::nullopt);                      
    // attr
    ifaContext.numHeads = num_query_heads;            
    ifaContext.preToken = pre_tokens;                 
    ifaContext.nextToken = next_tokens;               
    ifaContext.scaleValue = softmax_scale;            
    ifaContext.kvHeadNums = num_key_value_heads;      
    ifaContext.layOut = input_layout;                 
    ifaContext.blockSize = block_size;                
    ifaContext.innerPrecise = inner_precise;          
    ifaContext.antiquantMode = 0;                     
    ifaContext.softmaxLseFlag = return_softmax_lse;   
    ifaContext.keyAntiquantMode = key_quant_mode;     
    ifaContext.valueAntiquantMode = value_quant_mode; 
    ifaContext.sparseMode = sparse_mode;              
    ifaContext.queryQuantMode = query_quant_mode;     
    ifaContext.pseType = 0;                           
    ifaContext.windowSize = 0;                        
    // output
    ifaContext.attenOut = ToRequiredParaInfo(attenOut); 
    ifaContext.lseOut = ToRequiredParaInfo(lseOut);     

    ifaContext.kCache.resize(1);
    ifaContext.vCache.resize(1);
    at::IntArrayRef keyShapeRef = 
    ifaContext.kCache[0] = key.sizes();
    ifaContext.vCache[0] = value.sizes();
}

std::tuple<at::Tensor, at::Tensor> npu_fused_infer_attention_score_npu(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const c10::optional<at::Tensor> &query_rope, const c10::optional<at::Tensor> &key_rope,
    const c10::optional<at::Tensor> &pse_shift, const c10::optional<at::Tensor> &atten_mask,
    const c10::optional<at::Tensor> &actual_seq_qlen, const c10::optional<at::Tensor> &actual_seq_kvlen,
    const c10::optional<at::Tensor> &block_table, const c10::optional<at::Tensor> &dequant_scale_query,
    const c10::optional<at::Tensor> &dequant_scale_key, const c10::optional<at::Tensor> &dequant_offset_key,
    const c10::optional<at::Tensor> &dequant_scale_value, const c10::optional<at::Tensor> &dequant_offset_value,
    const c10::optional<at::Tensor> &dequant_scale_key_rope, const c10::optional<at::Tensor> &quant_scale_out,
    const c10::optional<at::Tensor> &quant_offset_out, const c10::optional<at::Tensor> &learnable_sink, 
    const c10::optional<at::Tensor> &metadata,
    int64_t num_query_heads, int64_t num_key_value_heads, double softmax_scale, int64_t pre_tokens, int64_t next_tokens,
    c10::string_view input_layout, int64_t sparse_mode, int64_t block_size, int64_t query_quant_mode,
    int64_t key_quant_mode, int64_t value_quant_mode, int64_t inner_precise, bool return_softmax_lse,
    c10::optional<int64_t> query_dtype, c10::optional<int64_t> key_dtype, c10::optional<int64_t> value_dtype,
    c10::optional<int64_t> query_rope_dtype, c10::optional<int64_t> key_rope_dtype,
    c10::optional<int64_t> key_shared_prefix_dtype, c10::optional<int64_t> value_shared_prefix_dtype,
    c10::optional<int64_t> dequant_scale_query_dtype, c10::optional<int64_t> dequant_scale_key_dtype,
    c10::optional<int64_t> dequant_scale_value_dtype, c10::optional<int64_t> dequant_scale_key_rope_dtype)
{
    std::string input_layout_str = std::string(input_layout);
    // construct the output tensor
    std::tuple<at::Tensor, at::Tensor> fia_output =
        construct_fia_output_tensor_v2(query, value, input_layout_str, quant_scale_out, block_table, num_query_heads,
                                       num_key_value_heads, return_softmax_lse, query_rope);
    at::Tensor output = std::get<0>(fia_output);
    at::Tensor softmax_lse = std::get<1>(fia_output);

    char input_layout_ptr[20];
    strncpy(input_layout_ptr, input_layout_str.c_str(), 20 - 1);

    IFAContext ifaContext;
    ConvertContextToParamsIFA(
        ifaContext, query, key, value, query_rope, key_rope, pse_shift, atten_mask, actual_seq_qlen, actual_seq_kvlen,
        block_table, dequant_scale_query, dequant_scale_key, dequant_offset_key, dequant_scale_value,
        dequant_offset_value, dequant_scale_key_rope, quant_scale_out, quant_offset_out, learnable_sink,
        num_query_heads, num_key_value_heads, softmax_scale, pre_tokens, next_tokens, input_layout_ptr, sparse_mode,
        block_size, query_quant_mode, key_quant_mode, value_quant_mode, inner_precise, return_softmax_lse, query_dtype,
        key_dtype, value_dtype, query_rope_dtype, key_rope_dtype, key_shared_prefix_dtype, value_shared_prefix_dtype,
        dequant_scale_query_dtype, dequant_scale_key_dtype, dequant_scale_value_dtype, dequant_scale_key_rope_dtype,
        output, softmax_lse);

    IFATiling ifaTiling;
    ifaTiling.DoSubOpTiling(ifaContext);
    // stream
    int devidx = query.device().index();
    c10_npu::NPUStream stream = c10_npu::getCurrentNPUStream(devidx);
    void *aclstream = stream.stream(false);
    // workspace
    auto workspaceTensor =
        at::empty({static_cast<long>(ifaContext.workSpaceSize)}, at::TensorOptions().dtype(at::kByte).device(query.options().device()));
    // tilingdata
    optiling::IncreFlashAttentionTilingData &tilingData = ifaContext.tilingData.tilingBase;
    uint8_t fdFlag = ifaContext.fdFlag;
    uint8_t layoutVal = ifaContext.layoutVal;
    uint8_t antiquantMode = ifaContext.antiquantMode_;
    // blockdim
    uint32_t blockDim = tilingData.increFlashAttentionSingleCoreParams.get_usedCoreNum();

    if (fdFlag == 0 && layoutVal == 0 && antiquantMode == 0) {
        LAUNCH_INCRE_FA(0, 0, 0);
    } else if (fdFlag == 0 && layoutVal == 0 && antiquantMode == 1) {
        LAUNCH_INCRE_FA(0, 0, 1);
    } else if (fdFlag == 0 && layoutVal == 1 && antiquantMode == 0) {
        LAUNCH_INCRE_FA(0, 1, 0);
    } else if (fdFlag == 0 && layoutVal == 1 && antiquantMode == 1) {
        LAUNCH_INCRE_FA(0, 1, 1);
    } else if (fdFlag == 1 && layoutVal == 0 && antiquantMode == 0) {
        LAUNCH_INCRE_FA(1, 0, 0);
    } else if (fdFlag == 1 && layoutVal == 0 && antiquantMode == 1) {
        LAUNCH_INCRE_FA(1, 0, 1);
    } else if (fdFlag == 1 && layoutVal == 1 && antiquantMode == 0) {
        LAUNCH_INCRE_FA(1, 1, 0);
    } else if (fdFlag == 1 && layoutVal == 1 && antiquantMode == 1) {
        LAUNCH_INCRE_FA(1, 1, 1);
    } else {
        printf("invalid flags: fd=%d layout=%d antiquantMode=%d\n",
            fdFlag, layoutVal, antiquantMode);
    }

    return std::tuple<at::Tensor, at::Tensor>(output, softmax_lse);
}

// 为META设备实现前向接口
std::tuple<at::Tensor, at::Tensor> npu_fused_infer_attention_score_meta(
    const at::Tensor &query, const at::Tensor &key, const at::Tensor &value,
    const c10::optional<at::Tensor> &query_rope, const c10::optional<at::Tensor> &key_rope,
    const c10::optional<at::Tensor> &pse_shift, const c10::optional<at::Tensor> &atten_mask,
    const c10::optional<at::Tensor> & actual_seq_qlen, const c10::optional<at::Tensor> & actual_seq_kvlen,
    const c10::optional<at::Tensor> &block_table, const c10::optional<at::Tensor> &dequant_scale_query,
    const c10::optional<at::Tensor> &dequant_scale_key, const c10::optional<at::Tensor> &dequant_offset_key,
    const c10::optional<at::Tensor> &dequant_scale_value, const c10::optional<at::Tensor> &dequant_offset_value,
    const c10::optional<at::Tensor> &dequant_scale_key_rope, const c10::optional<at::Tensor> &quant_scale_out,
    const c10::optional<at::Tensor> &quant_offset_out, const c10::optional<at::Tensor> &learnable_sink,
    const c10::optional<at::Tensor> &metadata,
    int64_t num_query_heads, int64_t num_key_value_heads, double softmax_scale, int64_t pre_tokens, int64_t next_tokens,
    c10::string_view input_layout, int64_t sparse_mode, int64_t block_size, int64_t query_quant_mode,
    int64_t key_quant_mode, int64_t value_quant_mode, int64_t inner_precise, bool return_softmax_lse,
    c10::optional<int64_t> query_dtype, c10::optional<int64_t> key_dtype, c10::optional<int64_t> value_dtype,
    c10::optional<int64_t> query_rope_dtype, c10::optional<int64_t> key_rope_dtype,
    c10::optional<int64_t> key_shared_prefix_dtype, c10::optional<int64_t> value_shared_prefix_dtype,
    c10::optional<int64_t> dequant_scale_query_dtype, c10::optional<int64_t> dequant_scale_key_dtype,
    c10::optional<int64_t> dequant_scale_value_dtype, c10::optional<int64_t> dequant_scale_key_rope_dtype)
{
    std::string input_layout_str = std::string(input_layout);
    // construct the output tensor
    std::tuple<at::Tensor, at::Tensor> fia_output =
        construct_fia_output_tensor_v2(query, value, input_layout_str, quant_scale_out, block_table, num_query_heads,
                                       num_key_value_heads, return_softmax_lse, query_rope);
    at::Tensor output = std::get<0>(fia_output);
    at::Tensor softmax_lse = std::get<1>(fia_output);
    return std::tuple<at::Tensor, at::Tensor>(output, softmax_lse);
}
} // namespace custom

// 为NPU设备注册前向实现
TORCH_LIBRARY_IMPL(custom, PrivateUse1, m)
{
    m.impl("npu_fused_infer_attention_score", &custom::npu_fused_infer_attention_score_npu);
}

// 为META设备注册前向实现
TORCH_LIBRARY_IMPL(custom, Meta, m)
{
    m.impl("npu_fused_infer_attention_score", &custom::npu_fused_infer_attention_score_meta);
}
